yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryProvider.scala

```diff
@@ -0,0 +1,1015 @@
+     */
+    def take(): Refresher.this.RefreshActions = {
+      val action = queue.take()
+      action
+    }
+
+    /**
+     * Flag to indicate the refresher thread is running
+     * @return
+     */
+    def isRunning(): Boolean  = {
+      running.get()
+    }
+
+    /**
+     * Get the last refresh time
+     * @return the last refresh time
+     */
+    def lastRefreshAttemptTime: Long = {
+      _lastRefreshAttemptTime.get()
+    }
+
+    /**
+     * Get count of messages processed.
+     * This will be at least equal to
+     * the number of refreshes executed
+     * @return
+     */
+    def messagesProcessed: Long = {
+      _messagesProcessed.get()
+    }
+
+    /**
+     * The number of actual refreshes triggered
+     * @return a count of refreshes
+     */
+    def refreshesExecuted: Long = {
+      _refreshesExecuted.get()
+    }
+
+    override def toString: String = {
+      s"Refresher running = $isRunning queue size = ${queue.size()};" +
+        s" processed = $messagesProcessed;" +
+        s" window = $refreshInterval mS;" +
+        s" refreshes executed = ${_refreshesExecuted.get()};" +
+        s" last refresh attempt = " + timeShort(lastRefreshAttemptTime, "never") + ";"
+    }
+  } // end class Refresher
+
+
+
+} // end class YarnHistoryProvider
+
+
+/**
+ * (Immutable) results of a list operation
+ * @param timestamp timestamp
+ * @param applications application listing. These must be pre-sorted
+ * @param failureCause exception raised (implies operation was a failure)
+ */
+private[spark] class ApplicationListingResults(
+    val timestamp: Long,
+    val applications: Seq[ApplicationHistoryInfo],
+    val failureCause: Option[Throwable]) {
+
+  /**
+   * Predicate which is true if the listing failed; that there
+   * is a failure cause value
+   * @return true if the listing failed
+   */
+  def failed: Boolean = { failureCause.isDefined }
+
+  def succeeded: Boolean = { !failed }
+
+  /**
+   * Get an updated time for display
+   * @return a date time or "never"
+   */
+  def updated: String = {
+    humanDateCurrentTZ(timestamp, YarnHistoryProvider.TEXT_NEVER_UPDATED)
+  }
+
+  /**
+   * Size of applications in the list
+   * @return
+   */
+  def size: Int = {
+    applications.size
+  }
+}
+
+/**
+ * Constants to go with the hstory provider.
+ *
+ * 1. Any with the prefix `KEY_` are for configuration (key, value) pairs, so can be
+ * searched for after scraping the History server web page.
+ *
+ * 2. Any with the prefix `OPTION_` are options from the configuration.
+ *
+ * 3. Any with the prefix `DEFAULT_` are the default value of options
+ *
+ * 4. Any with the prefix `TEXT_` are text messages which may appear in web pages
+ * and other messages (and so can be scanned for in tests)
+ */
+private[spark] object YarnHistoryProvider {
+
+  /**
+   * Default port. This is hard coded elsewhere in the history server; it is
+   * replicated here
+   */
+  val SPARK_HISTORY_UI_PORT_DEFAULT = 18080
```

Comment:
culled. I think I used it in test setup before moving to dynamic port assignment